aten/src/ATen/autocast_mode.h

```diff
@@ -3,17 +3,49 @@
 namespace at {
 namespace autocast {
 
-namespace {
-  bool is_autocast_eligible(const Tensor& tensor) {
-    return (tensor.is_cuda() || tensor.is_xla()) && tensor.is_floating_point();
-  }
-} // namespace
-
 TORCH_API bool is_enabled();
 TORCH_API void set_enabled(bool enabled);
 TORCH_API void clear_cache();
 TORCH_API int increment_nesting();
 TORCH_API int decrement_nesting();
+TORCH_API bool is_cpu_enabled();
+TORCH_API void set_cpu_enabled(bool enabled);
+TORCH_API at::ScalarType get_autocast_cpu_dtype();
+TORCH_API void set_autocast_cpu_dtype(at::ScalarType dtype);
+
+namespace {
+  bool is_autocast_eligible(const Tensor& tensor, DeviceType device_type) {
+    return device_type == DeviceType::CUDA
+        ? (tensor.is_cuda() || tensor.is_xla()) && tensor.is_floating_point()
+        : (tensor.is_cpu() || tensor.is_mkldnn()) && tensor.is_floating_point();
+  }
+} // namespace
+
+inline DispatchKey get_autocast_dispatch_key_from_device_type(
+    DeviceType device_type) {
+  switch (device_type) {
+    case DeviceType::CUDA:
+      return DispatchKey::Autocast;
+    case DeviceType::CPU:
+      return DispatchKey::AutocastCPU;
+    default:
+      throw std::runtime_error(
+          "unknown device type for autocast in get_autocast_dispatch_key_from_device_type");
+  }
+}
+
+inline at::ScalarType get_lower_precision_fp_from_device_type(
+    DeviceType device_type) {
+  switch (device_type) {
+    case DeviceType::CUDA:
+      return at::kHalf;
+    case DeviceType::CPU:
+      return get_autocast_cpu_dtype();
```

Comment:
If bfloat16 is the only dtype you support, are you sure you want to incur this TLS lookup in every autocast wrapper?

Reply:
Thank you for the review @mcarilli. Did you talk about the incur of `get_autocast_cpu_dtype()`? If so, I think the reason we keep it is for cases when different lower precision data types are supported. And when it happens, I think this way is the balance of code sharing and performance.